example/printer_raw.dart

```diff
@@ -0,0 +1,160 @@
+// Sends RAW data (string or hex sequences) directly to the printer
+
+import 'dart:convert';
+// ignore_for_file: depend_on_referenced_packages
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+import 'package:win32/win32.dart';
+
+class RawPrinter {
+  final String printerName;
+
+  RawPrinter(this.printerName);
+
+  /// Taken from console.dart example
+  /// Convert a Dart string to a heap stored byte array that can be passed through
+  /// FFI to an unmanaged API.
+  ///
+  /// The returned string is _not_ null-terminated.
+  Pointer<Uint8> _toCString(String buffer) {
+    final units = utf8.encode(buffer);
+    final result = calloc<Uint8>(units.length);
+    result.asTypedList(units.length).setAll(0, units);
+    return result;
+  }
+
+  Pointer<HANDLE> _startRawPrintJob(
+      {required String printerName,
+      required String documentTitle,
+      String dataType = 'RAW'}) {
+    final pPrinterName = printerName.toNativeUtf16();
+
+    final phPrinter = calloc<HANDLE>();
+    final pDocInfo = calloc<DOC_INFO_1>();
+
+    try {
+      // https://docs.microsoft.com/en-us/windows/win32/printdocs/openprinter
+      var fSuccess = OpenPrinter(pPrinterName, phPrinter, nullptr);
+      if (fSuccess == 0) {
+        final error = GetLastError();
+        throw Exception('OpenPrint error, status: $fSuccess, error: $error');
+      }
+
+      // https://docs.microsoft.com/en-us/windows/win32/printdocs/doc-info-1
+      pDocInfo.ref.pDocName = printerName.toNativeUtf16();
+      // We can use RAW or TEXT data types
+      pDocInfo.ref.pDatatype = TEXT(dataType); // RAW, TEXT or XPS_PASS
+      // "To print to a printer, set this to NULL."
+      pDocInfo.ref.pOutputFile = nullptr;
+
+      //https://docs.microsoft.com/en-us/windows/win32/printdocs/startdocprinter
+      fSuccess = StartDocPrinter(
+          phPrinter.value,
+          // "The version of the structure to which pDocInfo points. This value must be 1."
+          1,
+          pDocInfo);
+
+      if (fSuccess == 0) {
+        final error = GetLastError();
+        throw Exception(
+            'StartDocPrinter error, status: $fSuccess, error: $error');
+      }
+
+      return phPrinter;
+    } finally {
+      free(pPrinterName);
+      free(pDocInfo);
+    }
+  }
+
+  bool _startRawPrintPage(Pointer<HANDLE> phPrinter) {
+    //https://docs.microsoft.com/en-us/windows/win32/printdocs/startpageprinter
+    return StartPagePrinter(phPrinter.value) > 0;
+  }
+
+  bool _endRawPrintPage(Pointer<HANDLE> phPrinter) {
+    return EndPagePrinter(phPrinter.value) > 0;
+  }
+
+  bool _endRawPrintJob(Pointer<HANDLE> phPrinter) {
+    return EndDocPrinter(phPrinter.value) > 0 &&
+        ClosePrinter(phPrinter.value) > 0;
+  }
+
+  bool _printRawData(Pointer<HANDLE> phPrinter, String dataToPrint) {
+    final cWritten = calloc<DWORD>();
```

Comment:
Docs say that this returns a non-zero value for success, not that it's a positive integer.

Reply:
Nice catch, thanks.